File: src/main/java/net/snipsniper/sceditor/stamps/StampType.java
package net.snipsniper.sceditor.stamps;

import net.snipsniper.config.Config;
import net.snipsniper.sceditor.SCEditorWindow;

import java.util.HashMap;

public enum StampType {
    CUBE("Cube"),
    COUNTER("Counter"),
    CIRCLE("Circle"),
    SIMPLE_BRUSH("Simple Brush"),
    TEXT("Text"),
    RECTANGLE("Rectangle"),
    ERASER("Eraser");

    private static final HashMap<StampType, Integer> indexes = new HashMap<StampType, Integer>();
    private final String title;

    StampType(String title) {
        this.title = title;
    }

    //This way we can rearrange enums without worrying about updating their index manually
    public int getIndex() {
        if(indexes.containsKey(this))
            return indexes.get(this);

        int index = 0;
        for(StampType type : StampType.values()) {
            if (type == this) {
                indexes.put(this, index);
                return index;
            }
            index++;
        }
        return -1;
    }

    public String getTitle() {
        return title;
    }

    public IStamp getIStamp(Config config, SCEditorWindow scEditorWindow) {
        switch(this) {
            case CUBE: return new CubeStamp(config, scEditorWindow);
            case COUNTER: return new CounterStamp(config, scEditorWindow);
            case CIRCLE: return new CircleStamp(config, scEditorWindow);
            case SIMPLE_BRUSH: return new SimpleBrush(config, scEditorWindow);
            case TEXT: return new TextStamp(config, scEditorWindow);
            case RECTANGLE: return new RectangleStamp(config, scEditorWindow);
            case ERASER: return new EraserStamp(scEditorWindow, config);
        }
        return null;
    }

    public static StampType getByIndex(int index) {
        return StampType.values()[index];
    }

    public static int getSize() {
        return StampType.values().length;
    }
}

File: src/main/kotlin/net/snipsniper/sceditor/stamps/StampType.kt
package net.snipsniper.sceditor.stamps

import net.snipsniper.config.Config
import net.snipsniper.sceditor.SCEditorWindow

enum class StampType(val title: String, val iconFile: String) {
    CUBE("Marker", "marker"),
    COUNTER("Counter", "counter"),
    CIRCLE("Circle", "circle"),
    RECTANGLE("Rectangle", "rectangle"),
    SIMPLE_BRUSH("Simple Brush", "brush"),
    TEXT("Text", "text_tool"),
    ERASER("Eraser", "ratzefummel");

    //This way we can rearrange enums without worrying about updating their index manually
    val index: Int
        get() {
            if (indexCache.containsKey(this)) return indexCache[this]!!
            values().forEachIndexed { index, type ->
                if (type == this) return index
                indexCache[type] = index
            }
            return -1
        }

    fun getIStamp(config: Config, scEditorWindow: SCEditorWindow?) = when (this) {
        CUBE -> CubeStamp(config, scEditorWindow)
        COUNTER -> CounterStamp(config)
        CIRCLE -> CircleStamp(config)
        SIMPLE_BRUSH -> SimpleBrush(config, scEditorWindow)
        TEXT -> TextStamp(config, scEditorWindow)
        RECTANGLE -> RectangleStamp(config)
        ERASER -> EraserStamp(scEditorWindow, config)
    }

    companion object {
        private val indexCache = HashMap<StampType, Int>()
        fun getByIndex(index: Int) = values()[index]
        val size: Int get() = values().size
    }
}